docs/source/project_documentation/dev_guide/aac_constraints.md

```diff
@@ -0,0 +1,395 @@
+  a. Primitive constraints are declared within the primitive declaration, not at the point of usage.
+7. Once all of the parsed declarations are checked against declared constraints, any violations are reported as constraint errors.
+8. All constraint errors from all of the constraint rules are collected and returned together.
+  a. If there are no errors, the user is provided a success message
+  b. If there are errors, the definition fails the check.  There is an option on the `check` command to handle warnings as errors if desired.
+
+# Constraint Definitions
+In order for users to define the self-checking constraints in the AaC DSL, they have to declare constraint rules via the 3 types of constraint definitions. These definitions also provide contextual information for the constraint as well as behaviors and acceptance criteria -- something to leverage for automatically generating functional/integration tests in the future.
+
+## Generation File Safety
+
+We're going to discuss the use of `gen-plugin` below to help you build your constraints.  If you discover you need to update your plugin declaration, you can re-run `gen-plugin` without concern of losing any prior work.  If the output python files already exist, the generator will produce:
+
+  - `.aac_evaluate` files for any user-editable output with an already existing file.  You can look at the changes here and incorporate into your content.
+  - `.aac_backup` files for non-user editable output with an already existing file.  The backup contains the pre-existing content for your review.
+
+Once you're satisfied, run `aac clean <path_to_AaC_plugin_file> --code-output <src_path> --test-output <tests_path>` to remove any `.aac_evaluate` and `.aac_backup` files.
+
+Currently we are not generating documentation for plugins...so you're on your own for that.  Hopefully we'll get some content generation in place in the future to simplify plugin document content management.
+
+## Context Constraints
+Here is an example `context_constraint` definition:
+
+unique_root_keys.aac
+```yaml
+plugin:
+  name: Unique Root Keys
+  package: aac.plugins.unique_root_keys
+  description: An AaC plugin that enables validating all required fields are present in a definition.
+  context_constraints:
+    - name: Root key names are unique
+      description: Check every definition to ensure there are no duplicate root keys defined in the AaC language.
+```
+
+To implement the context constraint, run `aac gen-plugin <path-to-unique_root_keys.aac> --code-output <src-path> --test-output <tests-path>`.  Note that the package declaration will be used to establish sub-directories under your `--code-output` and `--test-output` paths.  By default, the `gen-plugin` command will prompt you to confirm the target output paths prior to attempting to generate any files.
```

Comment:
This is just a way to organize things...and as you say, it's not new and is well understood because it is used in broadly adopted development contexts.  I don't think it's unreasonable, odd, or non-pythonic to organize content in a package together.  I suggest we discuss pros/cons of this and alternatives...and based on that discussion make the appropriate change in a future minor release.